storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -88,85 +488,214 @@ type ReconcileVolume struct {
+			pv.Status.Reason, pv.Status.Message,
+		)
+		return delayedRequeue(err)
+	}
+	if _, err = r.setAvailableVolumeStatus(ctx, volume); err != nil {
+		return delayedRequeue(err)
+	}
+	reqLogger.Info("backing PersistentVolume is healthy")
+	return endReconciliation()
+}
+
+// Deploy a volume (i.e prepare the storage and create a PV).
+func (self *ReconcileVolume) deployVolume(
+	ctx context.Context,
+	request reconcile.Request,
+	volume *storagev1alpha1.Volume,
+) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+	nodeName := string(volume.Spec.NodeName)
 
-	// Check if this PV already exists
-	found := &corev1.PersistentVolume{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: pv.Name}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new PersistentVolume", "PersistentVolume.Name", pv.Name)
-		err = r.client.Create(context.TODO(), pv)
+	switch volume.Status.Job {
+	case "": // No job in progress: call Salt to prepare the volume.
+		if jid, err := self.salt.PrepareVolume(ctx, nodeName); err != nil {
+			reqLogger.Error(err, "failed to run PrepareVolume")
+			return delayedRequeue(err)
+		} else {
+			reqLogger.Info("start to prepare the volume")
+			self.recorder.Event(
+				volume, corev1.EventTypeNormal, "SaltCall",
+				"Salt call to PrepareVolume started",
+			)
+			return self.setPendingVolumeStatus(ctx, volume, jid)
+		}
+	case JOB_DONE_MARKER: // Salt job is done, now let's create the PV.
+		diskSize, err := self.getDiskSizeForVolume(ctx, volume)
 		if err != nil {
-			return reconcile.Result{}, err
+			reqLogger.Error(err, "call to GetVolumeSize failed")
+			return self.setFailedVolumeStatus(
+				ctx, volume, storagev1alpha1.CreationError,
+				"call to GetVolumeSize failed: %s", err.Error(),
+			)
+		}
+		pv := newPersistentVolume(volume, diskSize)
+		if err := self.createPersistentVolume(ctx, pv, volume); err != nil {
+			return delayedRequeue(err)
 		}
+		reqLogger.Info(
+			"creating a new PersistentVolume", "PersistentVolume.Name", pv.Name,
+		)
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "PvCreation",
+			"backing PersistentVolume created",
+		)
+		return self.setAvailableVolumeStatus(ctx, volume)
+	default: // PrepareVolume in progress: poll its state.
+		return self.pollSaltJob(
+			ctx, "PrepareVolume", volume, self.setPendingVolumeStatus,
+		)
+	}
+}
 
-		// PV created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
+// Finalize a volume marked for deletion.
+func (self *ReconcileVolume) finalizeVolume(
+	ctx context.Context,
+	request reconcile.Request,
+	volume *storagev1alpha1.Volume,
+) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+
+	if volume.Status.Phase == storagev1alpha1.VolumePending {
+		reqLogger.Info("pending volume cannot be marked for deletion: requeue")
```

Comment:
If we're here, the volume likely *is* 'marked for deletion', no?

Reply:
indeed, will reword that.